scripts/generate_unicode_data.cr

```diff
@@ -0,0 +1,161 @@
+
+record CaseRange, low : Int32, high : Int32, delta : Int32
+record AlternateRange, low : Int32, high : Int32
+record Stride, low : Int32, high : Int32, stride : Int32
+
+def case_ranges(entries, &block)
+  ranges = [] of CaseRange
+  first_codepoint = nil
+  last_codepoint = nil
+  first_match = nil
+  last_match = nil
+  entries.each do |entry|
+    codepoint = entry.codepoint
+    match = yield entry
+    if match
+      if last_codepoint == codepoint - 1 && last_match == match - 1
+        # Continue streak
+      else
+        if last_codepoint && last_match
+          ranges << CaseRange.new(first_codepoint.not_nil!, last_codepoint, first_match.not_nil! - first_codepoint.not_nil!)
+        end
+        first_codepoint = codepoint
+        first_match = match
+      end
+    else
+      if last_codepoint && last_match
+        ranges << CaseRange.new(first_codepoint.not_nil!, last_codepoint, first_match.not_nil! - first_codepoint.not_nil!)
+      end
+    end
+
+    last_codepoint = codepoint
+    last_match = match
+  end
+  ranges
+end
+
+def alternate_ranges(ranges)
+  alternate = [] of AlternateRange
+
+  first_codepoint = nil
+  last_codepoint = nil
+
+  ranges.each do |range|
+    codepoint = range.low
+    if last_codepoint == codepoint - 2
+      # Continue streak
+    else
+      if first_codepoint
+        alternate << AlternateRange.new(first_codepoint, last_codepoint.not_nil!)
+      end
+      first_codepoint = codepoint
+    end
+
+    last_codepoint = codepoint
+  end
+
+  if first_codepoint
+    alternate << AlternateRange.new(first_codepoint, last_codepoint.not_nil!)
+  end
+
+  alternate
+end
+
+def strides(entries, targets)
+  strides = [] of Stride
+
+  entries = entries.select { |entry| targets.includes?(yield entry) }
+
+  first_entry = nil
+  last_entry = nil
+  stride = nil
+
+  entries.each do |entry|
+    if first_entry
+      if last_entry
+        current_stride = entry.codepoint - last_entry.codepoint
+        if current_stride == stride
+          # Continue stride
+        else
+          if first_entry == last_entry
+            stride = current_stride
+          else
+            stride = 1 if first_entry.name.ends_with?("First>") && last_entry.name.ends_with?("Last>")
+            strides << Stride.new(first_entry.codepoint, last_entry.codepoint, stride.not_nil!)
+            first_entry = entry
+            stride = nil
+          end
+        end
+      end
+    else
+      first_entry = entry
+    end
+
+    last_entry = entry
+  end
+
+  if first_entry && last_entry
+    if stride
+      stride = 1 if first_entry.name.ends_with?("First>") && last_entry.name.ends_with?("Last>")
+      strides << Stride.new(first_entry.codepoint, last_entry.codepoint, stride)
+    else
+      strides << Stride.new(first_entry.codepoint, last_entry.codepoint, 1)
+    end
+  end
+
+  strides
+end
+
+entries = [] of Entry
+
+url = "http://www.unicode.org/Public/9.0.0/ucd/UnicodeData.txt"
+body = HTTP::Client.get(url).body
```

Comment:
Yes, this is generated once. Unicode data doesn't change very often so it's better here to generate it before compile-time. Generating it at compile-time, maybe with a macro, would be overkill (Elixir does that, though, but Elixir compiles modules to beam so they need to do it just once)